beacon_node/store/src/hot_cold_store.rs

```diff
@@ -562,15 +567,23 @@ impl<E: EthSpec> HotColdDB<E> {
         end_slot: Slot,
         end_block_hash: Hash256,
     ) -> Result<Vec<SignedBeaconBlock<E>>, Error> {
-        let mut blocks = ParentRootBlockIterator::new(self, end_block_hash)
-            .map(|(_, block)| block)
-            // Include the block at the end slot (if any), it needs to be
-            // replayed in order to construct the canonical state at `end_slot`.
-            .filter(|block| block.message.slot <= end_slot)
-            // Include the block at the start slot (if any). Whilst it doesn't need to be applied
-            // to the state, it contains a potentially useful state root.
-            .take_while(|block| block.message.slot >= start_slot)
-            .collect::<Vec<_>>();
+        let maybe_blocks: Result<Vec<SignedBeaconBlock<E>>, Error> =
+            ParentRootBlockIterator::new(self, end_block_hash)
+                .map(|result| result.map(|(_, block)| block))
+                // Include the block at the end slot (if any), it needs to be
+                // replayed in order to construct the canonical state at `end_slot`.
+                .filter(|result| match result {
```

Comment:
Nit: potentially a nice place for the recently-added [map_or](https://doc.rust-lang.org/std/result/enum.Result.html#method.map_or) function.

Reply:
Almost.  `filter` passes the iterator item to its closure by reference whereas `map_or` consumes `self`.

Reply:
You can use `result.as_ref().map_or(true, |block| ...)`

Reply:
You can always do `result.as_ref().map_or(..)` to avoid consuming self.